File: packages/core-rs/crates/mainframe-types/src/acp/jsonrpc.rs
//! JSON-RPC 2.0 envelope (ACP's transport framing). Method-specific params and
//! results (`InitializeRequest`, `ToolCallUpdate`, ...) are typed separately in
//! sibling modules and travel inside `params`/`result` as `serde_json::Value` —
//! the facade connection (group C, out of this task's scope) dispatches on
//! `method` before typing the payload.

use serde::{Deserialize, Serialize};
use serde_json::Value;

/// JSON-RPC request id: a number, a string, or (discouraged but legal) `null`.
/// `null` is represented by `Option<RequestId>::None`, which serializes to the
/// wire `null` on the envelope's `id` field without a `skip_serializing_if`.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(untagged)]
pub enum RequestId {
    Number(i64),
    Str(String),
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct JsonRpcRequest {
    pub jsonrpc: String,
    pub id: Option<RequestId>,
    pub method: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub params: Option<Value>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct JsonRpcNotification {
    pub jsonrpc: String,
    pub method: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub params: Option<Value>,
}

/// The wire type of a JSON-RPC/ACP error `code` — a plain `i32` per the
/// schema's `ErrorCode` (its "Other" arm permits any integer). Named
/// constants below cover the schema's predefined codes plus the one Mainframe
/// adds for unsupported-version negotiation (not an ACP-defined code — see
/// decisions).
pub type ErrorCode = i32;

pub mod error_codes {
    use super::ErrorCode;

    pub const PARSE_ERROR: ErrorCode = -32700;
    pub const INVALID_REQUEST: ErrorCode = -32600;
    pub const METHOD_NOT_FOUND: ErrorCode = -32601;
    pub const INVALID_PARAMS: ErrorCode = -32602;
    pub const INTERNAL_ERROR: ErrorCode = -32603;
    pub const REQUEST_CANCELLED: ErrorCode = -32800;
    pub const AUTHENTICATION_REQUIRED: ErrorCode = -32000;
    pub const RESOURCE_NOT_FOUND: ErrorCode = -32002;
    /// Mainframe-specific: `initialize` requested a `protocolVersion` the
    /// daemon does not support (spec acceptance criterion 1). Reserved-range
    /// (`-32000..-32099`) per the schema's guidance for protocol-specific codes.
    pub const UNSUPPORTED_PROTOCOL_VERSION: ErrorCode = -32001;
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct JsonRpcErrorObject {
    pub code: ErrorCode,
    pub message: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub data: Option<Value>,
}

/// The response body: exactly one of `result`/`error`, distinguished by which
/// key is present — the untagged variants try each shape in turn.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(untagged)]
pub enum JsonRpcOutcome {
    Result { result: Value },
    Error { error: JsonRpcErrorObject },
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct JsonRpcResponse {
    pub jsonrpc: String,
    pub id: Option<RequestId>,
    #[serde(flatten)]
    pub outcome: JsonRpcOutcome,
}

File: packages/types/src/acp/jsonrpc.ts
/**
 * JSON-RPC 2.0 envelope — ACP's transport framing (todo #350). Mirrors
 * `packages/core-rs/crates/mainframe-types/src/acp/jsonrpc.rs`. Method-
 * specific params/results are typed in the sibling modules and travel inside
 * `params`/`result` as `unknown` — the facade connection (out of this task's
 * scope) dispatches on `method` before typing the payload.
 */
import { z } from 'zod';

export const JsonRpcRequestIdSchema = z.union([z.number(), z.string()]);
export type JsonRpcRequestId = z.infer<typeof JsonRpcRequestIdSchema>;

export const JsonRpcRequestSchema = z
  .object({
    jsonrpc: z.literal('2.0'),
    id: JsonRpcRequestIdSchema.nullable(),
    method: z.string(),
    params: z.unknown().optional(),
  })
  .loose();
export type JsonRpcRequest = z.infer<typeof JsonRpcRequestSchema>;

export const JsonRpcNotificationSchema = z
  .object({
    jsonrpc: z.literal('2.0'),
    method: z.string(),
    params: z.unknown().optional(),
  })
  .loose();
export type JsonRpcNotification = z.infer<typeof JsonRpcNotificationSchema>;

export const JsonRpcErrorObjectSchema = z
  .object({
    code: z.number().int(),
    message: z.string(),
    data: z.unknown().optional(),
  })
  .loose();
export type JsonRpcErrorObject = z.infer<typeof JsonRpcErrorObjectSchema>;

/** Predefined JSON-RPC/ACP error codes; any other integer is a valid ACP-specific code. */
export const ErrorCodes = {
  PARSE_ERROR: -32700,
  INVALID_REQUEST: -32600,
  METHOD_NOT_FOUND: -32601,
  INVALID_PARAMS: -32602,
  INTERNAL_ERROR: -32603,
  REQUEST_CANCELLED: -32800,
  AUTHENTICATION_REQUIRED: -32000,
  RESOURCE_NOT_FOUND: -32002,
  /** Mainframe-specific: `initialize` requested an unsupported `protocolVersion` (acceptance criterion 1). */
  UNSUPPORTED_PROTOCOL_VERSION: -32001,
} as const;

const JsonRpcResponseResultSchema = z
  .object({
    jsonrpc: z.literal('2.0'),
    id: JsonRpcRequestIdSchema.nullable(),
    result: z.unknown(),
  })
  .loose();

const JsonRpcResponseErrorSchema = z
  .object({
    jsonrpc: z.literal('2.0'),
    id: JsonRpcRequestIdSchema.nullable(),
    error: JsonRpcErrorObjectSchema,
  })
  .loose();

/** The response body is exactly one of `result`/`error` — distinguished by which key is present. */
export const JsonRpcResponseSchema = z.union([JsonRpcResponseResultSchema, JsonRpcResponseErrorSchema]);
export type JsonRpcResponse = z.infer<typeof JsonRpcResponseSchema>;
